Tsp next 045

markcowl opened the pull request:
Splitting into smaller PRs that can be merged against 0.44.0:

- https://github.com/Azure/azure-rest-api-specs/pull/24297
- https://github.com/Azure/azure-rest-api-specs/pull/24303

mikeharder committed:
Set packages to latest/next

mikeharder committed:
Merge branch 'main' into typespec-next

mikeharder committed:
Merge branch 'main' into typespec-next

mikeharder committed:
Update typespec packages to dev versions

mikeharder committed:
Use "npm ci --force"

mikeharder committed:
Upgrade openapi and lint

mikeharder committed:
Remove package-lock.json and --force

mikeharder committed:
Use "next" package versions

mikeharder committed:
Add comment

mikeharder committed:
Merge branch 'main' into typespec-next

mikeharder committed:
Revert package-lock.json to match main

mikeharder committed:
Revert package-lock.json after install
- Prevents unintended "git diff"

markcowl committed:
Updating Contoso.WidgetManager

markcowl committed:
Update AnomalyDtector

markcowl committed:
Fixing ContentSafety

markcowl committed:
Fixes for healthinsights

markcowl committed:
Updating fleet.management

markcowl committed:
Fixing servicenetworking

markcowl committed:
Fixing Azure.AI.TextTranslation

mikeharder commented on specification/cognitiveservices/HealthInsights/healthinsights.common/primitives.tsp:
```diff
@@ -15,3 +17,5 @@ op LongRunningRpcOperation<
   Azure.Core.Foundations.ErrorResponse
 >;
+
+interface StandardOperations extends Azure.Core.ResourceOperations<NoClientRequestId & NoConditionalRequests & NoRepeatableRequests> {}
```
For my learning, is this considered a "breaking change"?

markcowl replied:
We should not need this change

mikeharder commented on specification/cognitiveservices/HealthInsights/healthinsights.oncophenotype/route.oncophenotype.tsp:
```diff
@@ -21,7 +22,8 @@ interface OncoPhenotype {
     "SuccessfulOncoPhenotypeGetAnalyzeStatus"
   )
-  getJob is Azure.Core.ResourceRead<OncoPhenotypeResult>;
+  getJob is StandardOperations.ResourceRead<OncoPhenotypeResult>;
```
For my learning, is this considered a "breaking change"?

markcowl replied:
We should not have any of the changes to use STandardOperations in this PR.  This was a removed breaking change

mikeharder commented on specification/contosowidgetmanager/Contoso.WidgetManager/main.tsp:
```diff
@@ -36,23 +38,24 @@ model Widget {
 interface Widgets {
   @doc("Fetch a Widget by name.")
-  getWidget is ResourceRead<Widget>;
+  getWidget is StandardOperations.ResourceRead<Widget>;
```
For my learning, is this considered a "breaking change"?

mikeharder replied:
Is this change compatible with both 44 and 45?  If so, I'd like to move this to a separate PR we merge to main before 45.

mikeharder commented on specification/contosowidgetmanager/data-plane/Azure.Contoso.WidgetManager/preview/2022-11-01-preview/widgets.json:
```diff
@@ -152,7 +152,7 @@
             "headers": {
               "Operation-Location": {
-                "description": "The location of an instance of OperationStatus",
+                "description": "The location for monitoring the operation state.",
```
Why did this description change, when I don't see a similar change in the tsp?

For my learning, is this considered a "breaking change"?

mikeharder replied:
Is this considered a "breaking change"?

markcowl replied:
No, field documentation would not be a breaking change under any definition I am aware of.

mikeharder commented on specification/contosowidgetmanager/data-plane/Azure.Contoso.WidgetManager/preview/2022-11-01-preview/widgets.json:
```diff
@@ -373,8 +373,7 @@
       "description": "The error object.",
       "required": [
         "code",
-        "message",
```
Why did this change, when I don't see a similar change in the tsp?

mikeharder commented on specification/contosowidgetmanager/data-plane/Azure.Contoso.WidgetManager/preview/2022-11-01-preview/widgets.json:
```diff
@@ -402,13 +401,11 @@
-      "required": [
-        "code"
-      ]
+      "description": "An object containing more specific information about the error. As per Microsoft One API guidelines - https://github.com/Microsoft/api-guidelines/blob/vNext/Guidelines.md#7102-error-condition-responses."
```
Why did this change?

mikeharder commented on specification/contosowidgetmanager/data-plane/Azure.Contoso.WidgetManager/preview/2022-11-01-preview/widgets.json:
```diff
@@ -417,7 +414,29 @@
         "name": "OperationState",
-        "modelAsString": true
+        "modelAsString": true,
+        "values": [
```
Why did this change?

mikeharder replied:
So this probably wouldn't be considered a "breaking change", because while it is a change to autorest output, it's not a "breaking" change to the autorest, since it's just documentation?

markcowl replied:
Yes, adding documentation is not a breaking change

mikeharder commented on specification/servicenetworking/ServiceNetworking.Management/main.tsp:
```diff
@@ -31,7 +31,7 @@ model Frontend is TrackedResource<FrontendProperties>{
   @visibility("read")
   @path
   @segment("frontends")
-  @pattern("[A-Za-z0-9]+[A-Za-z0-9-_.]{0,62}[A-Za-z0-9_]+")
```
I believe this is already updated in main from https://github.com/Azure/azure-rest-api-specs/pull/24225.  Maybe you need to merge latest main to your branch?

mikeharder replied:
This is fixed now that the PR is targeting typespec-next

mikeharder commented on specification/translation/Azure.AI.TextTranslation/examples/3.0/FindSentenceBoundaries_MinimumSet_Gen.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "title": "Find Sentence Boundaries",
+  "title": "Find Sentence Boundaries with minimum properties",
```
Why did this change?

mikeharder commented:
This is almost passing CI, just one more swagger with updated property order:

https://dev.azure.com/azure-sdk/public/_build/results?buildId=2818152&view=logs&j=9cd39ef8-47c1-5b58-08db-8c3708712f1b&t=8131bd26-22cb-54aa-fbbc-112785f95f08

mikeharder commented:
/azp run

mikeharder commented:
@markcowl: Are any of these changes compatible with both 44 and 45?  If so, I think we should move them to a separate PR to main, and merge before upgrading to 45.  This way, the PR upgrading to 45 will **only** contain reactions to truly breaking changes, and should be much smaller.

If you'd like, I can do the work of splitting the PR, if you can tell me which types of changes are and aren't compatible with 44.

markcowl commented:
@mikeharder That looks pretty close.  We are having a meeting to discuss breaking changes, so will revisit after we decide what to do with these.  One of them already has a fix checked in.

markcowl commented:
@mikeharder, we fixed one breaking change , reverted another, and decided to leave the other one as is, I will be updatign this as soon as the new changes are checked in and a new dev version is out

mikeharder commented on specification/cognitiveservices/AnomalyDetector/multivariate/models.tsp:
```diff
@@ -8,6 +8,7 @@ using Azure.Core;
 
 namespace AnomalyDetector.Multivariate;
 
+#suppress "@azure-tools/typespec-azure-core/documentation-required" "Legacy spec"
```
Why do you need to suppress `documentation-required` when there is `@doc` on the following line?  Is it because each enum value is also supposed to have `@doc`?

markcowl replied:
Yes, the scope of the suppression is for the entire enumeration

mikeharder commented:
/azp run

mikeharder commented:
/azp run specs - typespec - ci

markcowl commented on specification/cognitiveservices/HealthInsights/healthinsights.trialmatcher/route.trialmatcher.tsp:
```diff
@@ -21,7 +21,7 @@ interface TrialMatcher {
     "SuccessfulTrialMatcherGetAnalyzeStatus"
   )
-  getJob is Azure.Core.ResourceRead<TrialMatcherResult>;
+  getJob is StandardOperations.ResourceRead<TrialMatcherResult>;
```
I don't think we should have this change here.  This is a breaking change we removed

markcowl commented on specification/cognitiveservices/AnomalyDetector/main.tsp:
```diff
@@ -49,6 +49,8 @@ namespace AnomalyDetector;
 @doc("The secret key for your Azure Cognitive Services subscription.")
 model AnomalyDetectorApiKeyAuth is ApiKeyAuth<ApiKeyLocation.header, "Ocp-Apim-Subscription-Key">;
 
+@doc("The versions of this service")
```
This should just be a repeat of the documentation-required suppression, as in each of the cases below.  There is no value added with these doc comments (I did this one first before I saw how many fo these there were, and that I would not have sensible documentation comments for all of them)

markcowl replied:
This is a nit, though, the doc comments here also resolve the issue

mikeharder replied:
Fixed in https://github.com/Azure/azure-rest-api-specs/pull/24297

markcowl commented on specification/contosowidgetmanager/Contoso.WidgetManager/main.tsp:
```diff
@@ -16,8 +16,10 @@ using Azure.Core;
 
 namespace Azure.Contoso.WidgetManager;
 
+@doc("The service")
```
This one should also be just a suppression, liek the others

markcowl commented on specification/contosowidgetmanager/Contoso.WidgetManager/main.tsp:
```diff
@@ -36,23 +38,24 @@ model Widget {
   sharedModel?: FakedSharedModel;
 }
 
+interface StandardOperations extends Azure.Core.ResourceOperations<Azure.Core.Traits.NoClientRequestId & Azure.Core.Traits.NoConditionalRequests & Azure.Core.Traits.NoRepeatableRequests, Azure.Core.Foundations.ErrorResponse> {}
```
We do not need this change

markcowl commented on specification/contosowidgetmanager/Contoso.WidgetManager/main.tsp:
```diff
@@ -36,23 +38,24 @@ model Widget {
 interface Widgets {
   @doc("Fetch a Widget by name.")
-  getWidget is ResourceRead<Widget>;
+  getWidget is StandardOperations.ResourceRead<Widget>;
```
let's remove this change, this is one of the breaking changes that was removed

markcowl commented on specification/contosowidgetmanager/Contoso.WidgetManager/main.tsp:
```diff
@@ -36,23 +38,24 @@ model Widget {
 
   @doc("Gets status of a Widget operation.")
-  getWidgetOperationStatus is GetResourceOperationStatus<Widget>;
+  getWidgetOperationStatus is StandardOperations.GetResourceOperationStatus<Widget>;
```
let's remove this change, this is one of the breaking changes that was removed

markcowl commented on specification/contosowidgetmanager/Contoso.WidgetManager/main.tsp:
```diff
@@ -36,23 +38,24 @@ model Widget {
   @doc("Creates or updates a Widget asynchronously.")
   @pollingOperation(Widgets.getWidgetOperationStatus)
-  createOrUpdateWidget is LongRunningResourceCreateOrUpdate<Widget>;
+  createOrUpdateWidget is StandardOperations.LongRunningResourceCreateOrUpdate<Widget>;
```
let's remove this change, this is one of the breaking changes that was removed

markcowl commented on specification/contosowidgetmanager/Contoso.WidgetManager/main.tsp:
```diff
@@ -36,23 +38,24 @@ model Widget {
   @doc("Delete a Widget asynchronously.")
   @pollingOperation(Widgets.getWidgetOperationStatus)
-  deleteWidget is LongRunningResourceDelete<Widget>;
+  deleteWidget is StandardOperations.LongRunningResourceDelete<Widget>;
```
let's remove this change, this is one of the breaking changes that was removed

markcowl commented on specification/contosowidgetmanager/Contoso.WidgetManager/main.tsp:
```diff
@@ -36,23 +38,24 @@ model Widget {
 
   @doc("List Widget resources")
-  listWidgets is ResourceList<
+  listWidgets is StandardOperations.ResourceList<
```
let's remove this change, this is one of the breaking changes that was removed

markcowl committed:
Update helpers.tsp for object deprecation

markcowl commented:
/azp run specs - typespec - ci

mikeharder commented on specification/cognitiveservices/data-plane/HealthInsights/preview/2023-03-01-preview/openapi.json:
```diff
@@ -78,7 +78,8 @@
               "$ref": "#/definitions/OncoPhenotypeData"
-            }
+            },
+            "required": true
```
@markcowl: What is causing the addition of "required" here?  Is this a breaking change?
